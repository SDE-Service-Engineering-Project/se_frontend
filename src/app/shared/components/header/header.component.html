<nav class="navbar navbar-expand-lg bg-primary" data-bs-theme="dark">
  <div class="container-fluid">
    <a class="navbar-brand" routerLink="/home">Home</a>
<button class="navbar-toggler" type="button" data-testid="toggle-btn" (click)="isNavbarCollapsed = !isNavbarCollapsed" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
  <span class="navbar-toggler-icon"></span>
</button>
    <div [ngbCollapse]="isNavbarCollapsed" class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
        <li class="nav-item" *ngFor="let item of navItems">
          <a class="nav-link" [attr.data-testid]="item.testId" routerLinkActive="active-link"
             [routerLink]="item.path">{{item.title}}</a>
        </li>
      </ul>
      <ul class="navbar-nav">
        <li class="nav-item">
          <a id="logout" class="nav-link" data-testid="nav-logout" (click)="logout()" >Logout</a>
        </li>
      </ul>
    </div>
  </div>
</nav>
