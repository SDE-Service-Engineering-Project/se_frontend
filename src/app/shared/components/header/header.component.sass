.navbar-brand
  color: white

.nav-link
  color: white

.navbar-nav
  display: flex

#logout
  float: right
  cursor: pointer
