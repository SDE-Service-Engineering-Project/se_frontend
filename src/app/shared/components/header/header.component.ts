import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { StorageService } from '../../../services/storage.service';

@Component({
  selector: 'app-header',
  templateUrl: './header.component.html',
  styleUrls: ['./header.component.sass'],
})
export class HeaderComponent {
  isNavbarCollapsed = false;
  navItems = [
    {
      title: 'Catalog',
      path: '/catalog',
      testId: 'nav-item-catalog',
    },
    {
      title: 'Rented Cars',
      path: '/cars',
      testId: 'nav-item-cars',
    },
    {
      title: 'Find a Location',
      path: '/maps',
      testId: 'nav-item-maps',
    },
  ];

  constructor(private storageService: StorageService, private router: Router) {}

  logout(): void {
    this.storageService.removeAll();
    this.router.navigate(['/']);
  }
}
