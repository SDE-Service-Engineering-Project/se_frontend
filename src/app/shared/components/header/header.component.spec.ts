import { ComponentFixture } from '@angular/core/testing';
import { Location } from '@angular/common';

import { HeaderComponent } from './header.component';
import {
  byTestId,
  createComponentFactory,
  Spectator,
} from '@ngneat/spectator/jest';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbCollapseModule } from '@ng-bootstrap/ng-bootstrap';
import { Component } from '@angular/core';
import { Router } from '@angular/router';

@Component({
  template: '',
})
class RoutedTestComponent {}

describe('HeaderComponent', () => {
  let component: HeaderComponent;
  let fixture: ComponentFixture<HeaderComponent>;
  let spectator: Spectator<HeaderComponent>;

  const createComponent = createComponentFactory({
    component: HeaderComponent,
    imports: [
      NgbCollapseModule,
      RouterTestingModule.withRoutes([
        {
          path: 'cars',
          component: RoutedTestComponent,
        },
        {
          path: 'maps',
          component: RoutedTestComponent,
        },
        {
          path: 'catalog',
          component: RoutedTestComponent,
        },
        {
          path: '',
          component: RoutedTestComponent,
        },
      ]),
    ],
  });

  beforeEach(() => {
    spectator = createComponent();
    fixture = spectator.fixture;
    component = spectator.component;
  });

  it('should render the component', () => {
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });

  it('should navigate to the correct routes', () => {
    spectator.click(getNavigationItem('cars'));
    spectator.detectChanges();
    expect(spectator.inject(Location).path()).toBe('/cars');
    spectator.click(getNavigationItem('maps'));
    spectator.detectChanges();
    expect(spectator.inject(Location).path()).toBe('/maps');
    spectator.click(getNavigationItem('catalog'));
    spectator.detectChanges();
    expect(spectator.inject(Location).path()).toBe('/catalog');
  });

  //TODO: add expectation of real logout i.e. token was removed, etc.
  it('should log the user out', () => {
    spectator.click(getLogoutNavItem());
    expect(spectator.inject(Location).path()).toBe('/');
  });

  it('should change the collapsed property of the navbar', () => {
    spectator.click(getNavToggleButton());
    expect(component.isNavbarCollapsed).toEqual(true);
  });

  function getNavigationItem(title: string): HTMLButtonElement {
    return spectator.query(byTestId(`nav-item-${title}`)) as HTMLButtonElement;
  }

  function getLogoutNavItem(): HTMLButtonElement {
    return spectator.query(byTestId('nav-logout')) as HTMLButtonElement;
  }

  function getNavToggleButton(): HTMLButtonElement {
    return spectator.query(byTestId('toggle-btn')) as HTMLButtonElement;
  }
});
