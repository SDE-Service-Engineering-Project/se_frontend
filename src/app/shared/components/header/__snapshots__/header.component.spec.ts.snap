// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderComponent should render the component 1`] = `
<div
  id="root0"
>
  <nav
    class="navbar navbar-expand-lg bg-primary"
    data-bs-theme="dark"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/home"
        routerlink="/home"
      >
        Navbar
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-bs-target="#navbarSupportedContent"
        data-bs-toggle="collapse"
        data-testid="toggle-btn"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse show"
        id="navbarSupportedContent"
      >
        <ul
          class="navbar-nav me-auto mb-2 mb-lg-0"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              data-testid="nav-item-cars"
              href="/cars"
              routerlinkactive="active-link"
            >
              Rented Cars
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              data-testid="nav-item-maps"
              href="/maps"
              routerlinkactive="active-link"
            >
              Find a Location
            </a>
          </li>
          
        </ul>
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              data-testid="nav-logout"
              id="logout"
            >
              Logout
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
